rfcs/0112-demote-darwin-support-tier.md

```diff
@@ -0,0 +1,91 @@
+    - Due to Hydra being understaffed, this is a problem (see above).
+
+We are not only lacking CI machines, but also maintainers that can dedicate their time to work on fixing broken Darwin packages.
+
+## Exemplary collection of issues with Darwin
+
+The issues that are constantly encountered in relation to the Darwin platform have a genuine negative impact on the `nixpkgs` development experience. People are forced to deal with these problems even though they have no hardware to test on, lack the proper expertise, and are reluctant when it comes to investing additional time and energy to patch for the failing platform. Demoting Darwin to Tier 3 will allow more developers to refocus their attention on better supported platforms, care less about Darwin issues, and push that burden onto the Darwin maintainers.
+
+An excerpt of these issues are:
+
+- The only way to debug issues without owning a Mac is to run Darwin in a VM, of which the legal situation is unclear.
+   - Those licensing requirements are causing friction with our CI setup as well.
+- The sandboxing capabilities on MacOS are limited. This results in a special category of test failures only relevant to Darwin. It also prevents us from providing a community builder, because it could not be shared between multiple people that cannot trust each other.
+- The MacOS SDK cannot be updated and is stuck on 10.12 (released 2016/09) because Apple does not publish the required sources.
+    - https://github.com/NixOS/nixpkgs/issues/101229#issuecomment-938747052
+    - It is highly unlikely that we can report bugs to projects with such an outdated and unsupported SDK version
+- The MacOS SDK being outdated blocks nixpkgs moving to go 1.17:
+    - https://github.com/NixOS/nixpkgs/pull/127519#issuecomment-864926149
+    - This will become a growing pain point once more packages become go 1.17-only (tailscale, talosctl, …)
+- Multiple Packages were marked as broken, because they require symbols from newer MacOS SDK versions.
+    - https://github.com/NixOS/nixpkgs/commit/3ceb5ab5ed0d1fcbe53ef00621f44d61dc524796
+    - https://github.com/NixOS/nixpkgs/commit/c9a3ac5d3cb5e910238d01e534d74d5d50e4b6b7
+- Curl added a SystemConfiguration dependency for NAT64 support, which introduced a reference loop, requiring a downstream patch to workaround.
+    - This blocked updating to more recent curl versions for most of the last release cycle.
+    - https://github.com/NixOS/nixpkgs/pull/124502#issuecomment-850834981
+- Enabling brotli support by default in curl broke the Darwin stdenv, which draws in a great number of packages.
+    - https://github.com/NixOS/nixpkgs/pull/112947
+    - https://github.com/NixOS/nixpkgs/pull/115498
+
+# Detailed design
+[design]: #detailed-design
+
+- Together with this RFC, `0046-platform-support-tiers.md` will be updated accordingly (simply move `x86_64-darwin` down one section).
+- Whatever needs to be done to make Darwin not block any channels anymore (TODO).
```

Comment:
> Something I'm missing is a list of concrete requirements to reinstate x86_64-darwin to tier 2. […] Can we quantify what would need to happen?

The argument is "Darwin does not meet the Tier 2 requirements right now", so the requirements to go back to Tier 2 are laid out in RFC 46.

> Is there a specific number of packages we can aim to support or have I misunderstood and this point isn't actually about package availability?

I think there are two aspects to the "most packages work". The one is the user perspective, i.e. how many packages are available in the repository and not marked as broken. I'll trust your judgement on that. But the other one is the developer perspective, i.e. how often do packages break in Darwin-specific ways. Which goes back to your point mentioned above:

> Maintainers have reasons to be wary but "No extra support burden," isn't a realistic goal.

- There seems to be a general consensus that the current extra support burden is "too much"
- Note that due to the nature of the platform (proprietary, further away from Linux, etc.), expect the amount of friction developers will be accepting to put up with to be lower than for example with `aarch64-linux`.

Reply:
I think this comes back to definining a set of e.g. 100 core packages, that should be maintained at all times, in addition to the tiers. Tier 1 and 2 should quickly fix any breakage of the core packages.

Reply:
I'm asking for clarification of "Darwin does not meet the Tier 2 requirements." RFC #46 isn't very normative and I don't believe that was even the intent behind it.

These are the tier 2 requirements, as I read RFC 46:

- A lot of packages built by Hydra, full ofBorg support.
- Most packages work, credible ambition to reach Tier 1 coverage at some point.

> A lot of packages built by Hydra,

Can we put a number on this? Say, 75% of x86_64-Linux packages?

> full ofBorg support.

What does this mean, ofBorg supports Darwin builds, so that's not enough? Can we put a time on this, like within 24h? Sidenote: Can we change maintainer policy, temporarily, to avoid having patches applied across all platforms, excluding Darwin if the patch isn't needed there and would cause a stdenv rebuild, To avoid overpowering the CI capacity we do have.

> Most packages work,

Can we put a number on this, like "less than 10% of packages are marked broken?"

> credible ambition to reach Tier 1 coverage at some point.

How can we, as Darwin maintainers, demonstrate whether there is "credible ambition" to reach tier 1 and how far in the future can "some point" be?

---

I agree the current burden indeed seems like too much, only the core maintainers can really judge this and I don't see many voices in opposition of the sentiment. That's why I'm not arguing against demoting Darwin (though I do feel like grouping x86_64-darwin with aarch64-darwin feels overly harsh). What I am asking for is a clear way back to tier 2. If any maintainer can just shoot down attempts to reinstate Darwin as a tier 2 platform on the basis of the negative sentiment already incurred by the current situation with unclear statements like "it doesn't meet RFC 46 requirements," then how can we be expected to work towards this and not get demotivated?

It's very unfortunate that the crux of this issue (CI/ofBorg) is precisely something the community of Darwin users is relatively powerless to do much about.

Reply:
> I'm asking for clarification of "Darwin does not meet the Tier 2 requirements." RFC #46 isn't very normative and I don't believe that was even the intent behind it.

> It's very unfortunate that the crux of this issue (CI/ofBorg) is precisely something the community of Darwin users is relatively powerless to do much about.

Unfortunately, these two things are in fact related: as Tier 2 includes, among other things, commitments around maintaining the capacity for Hydra/OfBorg, Tier 2 and Tier 1 are not questions of a purely technical status review, but also of organisational coordination; RFC 46 intentionally did not try to predict how access and, if applicable, financial issues will be negotiated…

Reply:
> though I do feel like grouping x86_64-darwin with aarch64-darwin feels overly harsh

Yeah, I'm a bit sorry about that. When I wrote the RFC I still thought `aarch64-darwin` was in Tier 7 lol. The problem there seems to be that hydra building packages is bound to the support tier right now. We could change that and move `aarch64-darwin` to Tier 4, or create a new tier in between or create new rules or something. But in the end I had to deem this out of scope for this specific RFC.

Generally, this discussion has shown us a lot of flaws in our current support tier handling. I would be in favor of an RFC that improves that.

> It's very unfortunate that the crux of this issue (CI/ofBorg) is precisely something the community of Darwin users is relatively powerless to do much about.

Yes, it really is. I hope that at the very least, this whole discussion gives us new energy to bring these issues forward.